### PR TITLE
LoadILLDiffraction fix detector dimension size

### LINKED ISSUE #20105: Fix LoadILLDiffraction for D20 case.
Regression is introduced in PR #19920, that broke the loading for D20, which has no tubes, just a 1D detector array.

### PR DESCRIPTION
**Pull request by gvardany**
This PR fixes the regression introduced in PR #19920 , where the sizes of the data matrix dimensions got confused/swapped.

It was not caught, since ironically D2B has 128 tubes, and 128 pixels in each tube.

Whereas for D20, the scanning workspace case had no unit test, since the building of a scanning workspace currently takes too much time (>40sec). 

This will be addressed in ScanningWorkspaceBuilder in a separate PR, meanwhile I have added a system test instead.

Another issue is fixed with absolute times, there was a superfluous sec to nano sec conversion.

The unit tests are hence made more rigorous, the documentation is also improved.

**To test:**

Unit and system tests pass.

Fixes #20105 . 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gvardany** [comment]
tagging @ianbush to stay in touch

File: Framework/DataHandling/src/LoadILLDiffraction.cpp [hunk 1]
```diff
@@ -554,13 +553,13 @@ LoadILLDiffraction::getDurations(const NXDouble &scan) const {
...
 LoadILLDiffraction::getAbsoluteTimes(const NXDouble &scan) const {
-  std::vector<DateAndTime> times; // in units of ns
+  std::vector<DateAndTime> times;
   std::vector<double> durations = getDurations(scan);
   DateAndTime time = m_startTime;
   times.emplace_back(time);
   size_t timeIndex = 1;
   while (timeIndex < m_numberScanPoints) {
-    time += durations[timeIndex - 1] * 1E9; // from sec to ns
+    time += durations[timeIndex - 1];
```
Comment (ianbush) [inline Framework/DataHandling/src/LoadILLDiffraction.cpp]: :+1: as durations is a vector of doubles this seems correct.

File: Framework/DataHandling/inc/MantidDataHandling/LoadILLDiffraction.h [hunk 2]
```diff
@@ -96,9 +96,10 @@ class MANTID_DATAHANDLING_DLL LoadILLDiffraction
   void resolveInstrument();
   void resolveScanType();
 
-  size_t m_sizeDim1;            ///< size of dim1, either tubes or detectors
-  size_t m_sizeDim2;            ///< size of dim2, used for pixels within tubes
-  size_t m_numberDetectorsRead; ///< number of cells read from file
+  size_t m_sizeDim1; ///< size of dim1, number of pixels in a tube (D2B) or the
+                     /// whole detector (D20)
+  size_t m_sizeDim2; ///< size of dim2, number of tubes (1 for D20!)
```
Comment (ianbush) [inline Framework/DataHandling/inc/MantidDataHandling/LoadILLDiffraction.h]: I think this is back to front, `m_sizeDim1` should be number of tubes and `m_sizeDim2` number of pixels for D2B.

File: Framework/DataHandling/src/LoadILLDiffraction.cpp [hunk 3]
```diff
@@ -391,7 +391,7 @@ void LoadILLDiffraction::fillStaticInstrumentScan(const NXUInt &data,
        i < m_numberDetectorsActual + NUMBER_MONITORS; ++i) {
     auto &spectrum = m_outWorkspace->mutableY(i);
     auto &errors = m_outWorkspace->mutableE(i);
-    const auto tubeNumber = (i - NUMBER_MONITORS + deadOffset) / m_sizeDim1;
+    const auto tubeNumber = (i - NUMBER_MONITORS + deadOffset) / m_sizeDim2;
```
Comment (ianbush) [inline Framework/DataHandling/src/LoadILLDiffraction.cpp]: I think this change is correct.

File: docs/source/release/v3.11.0/framework.rst [hunk 4]
```diff
@@ -45,6 +45,7 @@ Bug fixes
 #########
 
 - :ref:`CubicSpline <func-CubicSpline>` is fixed to sort the y-values and x-values correctly.
+- A bug in :ref:`LoadILLDiffraction <algm-LoadILLDiffraction-v1>`, concerning the dimensions of the 1D or 2D detector data, has been fixed.
```
Comment (ianbush) [inline docs/source/release/v3.11.0/framework.rst]: I don't think this is required as this bug never made it into a release.

File: docs/source/algorithms/LoadILLDiffraction-v1.rst [hunk 5]
```diff
@@ -10,15 +10,39 @@
...
-In this case there is only one bin, and a workspace is produced with time indexed detector information.
+For D20 1-dimensional detector, it supports 3 resolution modes:
+
+- **low** is the physical configuration, i.e. 1536 pixels. This corresponds to the IDF *D20_lr*.
+
+- **nominal** is the default configuration, i.e. pixels split by 2 by DAQ, so in total 3072 pixels. This corresponds to IDF *D20*.
+
+- **high** is when each pixel is split by 3, resulting in 4608 pixels. This corresponds to IDF *D20_hr*.
+
+Note, that all the IDFs contain only active pixels, and do not count the first and last banks went dead recently.
```
Comment (ianbush) [inline docs/source/algorithms/LoadILLDiffraction-v1.rst]: I would say `first and last banks which are inactive.`, as recently will be misleading in a years time.

**Review by ianbush — COMMENTED** [review]
Changes and addition of test look good to me. My main question is the changing around of the comments - we can discuss when you return.

**Review by ianbush — APPROVED** [review]
**Gatekeeper** - My only comments were changes to code comments and documentation. As it would be useful to get this merged, and @gvardany is away for a couple more weeks, I pushed the changes myself.

**Pull request merged**